### PR TITLE
Refactor: respond to client even when leader is switched

### PR DESCRIPTION
**Pull request by drmingdrmer**
##### Refactor: respond to client even when leader is switched

Before this commit, the channels used for sending responses to the
client were dropped whenever a leader stepped down, which was not
necessary.

In the current commit, when a leader quits, we now preserve the channel, ensuring that:

- An OK response is sent to the client if the log entry the client has
  written is successfully committed;

- Alternatively, a `ForwardToLeader` error is sent if the log is
  reverted due to a conflict with the new leader.

In short, OpenRaft now makes a concerted effort to deliver a success
message to the client wherever possible.

Two test cases have been added to cover both successful and erroneous
scenarios.

Special thanks to @tvsfx for proposing this refinement :D

- Fix: #963

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/973)

### CODE REVIEW

**Review by drmingdrmer — COMMENTED** [review]
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/973)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @ariesdevil and @lichuang)

**Review by tvsfx — COMMENTED** [review]
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/973)* status: all files reviewed, 1 unresolved discussion (waiting on @ariesdevil, @drmingdrmer, and @lichuang)

___
*[`openraft/src/core/raft_core.rs` line 1582 at r1](https://reviewable.io/reviews/datafuselabs/openraft/973#-Nm044hY8qX1LYsR5mfN:-Nm044hY8qX1LYsR5mfO:bpwjnlm) ([raw file](https://github.com/datafuselabs/openraft/blob/ea017289d9f30f9cc40d5f1be6e5d20e44d3769a/openraft/src/core/raft_core.rs#L1582)):*
> ```XML
>                     AsyncRuntimeOf::<C>::spawn(async move {
>                         for (log_index, tx) in removed.into_iter() {
>                             let res = tx.send(Err(ClientWriteError::ForwardToLeader(ForwardToLeader {
> ```

This is actually a nice extra benefit of this approach that I did not think of; rather than having an empty `ForwardToLeader` error like before in `QuitLeader`, you can now redirect the client to the new leader immediately.
